File: ios/RCTIONImage2DContext.m
//
//  RCTIONImage2DContext.m
//  RCTIONImage2D
//
//  Created by Marko on 26/12/15.
//
//

#import "RCTIONImage2DContext.h"
#import "UIColor+ParseColor.h"
#import <ImageIO/ImageIO.h>
#import <MobileCoreServices/MobileCoreServices.h>

@implementation RCTIONImage2DContext {
    NSString *_contextId;
    CGImageRef _image;
}

- (id)init {
    self = [super init];
    if (self != nil) {
        _contextId = [[NSUUID UUID] UUIDString];
    }
    return self;
}

- (NSString *)getContextId {
    return _contextId;
}

- (void)createFromFileUrl:(NSString *)fileUrl
             withMaxWidth:(NSInteger)maxWidth
            withMaxHeight:(NSInteger)maxHeight {
    NSData *imageData = [[NSData alloc] initWithContentsOfURL:[NSURL URLWithString:fileUrl]];
    [self createFromData:imageData
             withMaxWidth:maxWidth
            withMaxHeight:maxHeight];
}

- (void)createFromBase64String:(NSString*)base64
                  withMaxWidth:(NSInteger)maxWidth
                 withMaxHeight:(NSInteger)maxHeight {
    NSData *imageData = [[NSData alloc] initWithBase64EncodedString:base64
                                                            options:0];
    [self createFromData:imageData
            withMaxWidth:maxWidth
           withMaxHeight:maxHeight];
}

- (void)createFromData:(NSData*)imageData
          withMaxWidth:(NSInteger)maxWidth
         withMaxHeight:(NSInteger)maxHeight {
    CGImageSourceRef imageSource = CGImageSourceCreateWithData((__bridge CFDataRef)imageData, NULL);
    CGImageRef newImage = CGImageSourceCreateImageAtIndex(imageSource, 0, NULL);
    CFRelease(imageSource);
    
    [self createFromImage:newImage
            withMaxWidth:maxWidth
           withMaxHeight:maxHeight];
    
    CGImageRelease(newImage);
}

- (void)createFromImage:(CGImageRef)image
           withMaxWidth:(NSInteger)maxWidth
          withMaxHeight:(NSInteger)maxHeight {
    double orgWidth = CGImageGetWidth(image);
    double orgHeight = CGImageGetHeight(image);
    
    if (orgWidth <= maxWidth && orgHeight <= maxHeight) {
        _image = CGImageRetain(image);
        return;
    }
    
    double newWidth = maxWidth;
    double newHeight = maxHeight;
    
    if (orgWidth / maxWidth < orgHeight / maxHeight) {
        newWidth = (newHeight / orgHeight) * orgWidth;
    } else {
        newHeight = (newWidth / orgWidth) * orgHeight;
    }
    
    CGColorSpaceRef colorSpace = CGImageGetColorSpace(image);
    CGContextRef context = CGBitmapContextCreate(nil,
                                                 newWidth,
                                                 newHeight,
                                                 CGImageGetBitsPerComponent(image),
                                                 32 * newWidth,
                                                 colorSpace,
                                                 CGImageGetAlphaInfo(image));
    CGContextDrawImage(context, CGRectMake(0, 0, newWidth, newHeight), image);
    
    _image = CGBitmapContextCreateImage(context);
    
    CGContextRelease(context);
}

- (NSString *)save:(NSString *)fileName {
    NSData *imageData = [self getAsData];
    
    NSArray *paths = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES);
    NSString *documentsDirectory = [paths firstObject];
    NSFileManager *fileManager = [NSFileManager defaultManager];
    
    NSString *imagePath = [[documentsDirectory stringByAppendingPathComponent:fileName] stringByAppendingPathExtension:@"jpg"];
    
    [fileManager createFileAtPath:imagePath contents:imageData attributes:nil];
    
    return imagePath;
}

- (NSString *)getAsBase64String {
    return [[self getAsData] base64EncodedStringWithOptions:0];
}

- (NSData *)getAsData {
    NSDictionary *properties;
    CFMutableDataRef imageData = CFDataCreateMutable(NULL, 0);
    CFStringRef type = UTTypeCreatePreferredIdentifierForTag(kUTTagClassMIMEType, (__bridge CFStringRef) @"image/png", kUTTypeImage);
    CGImageDestinationRef destination = CGImageDestinationCreateWithData(imageData, type, 1, NULL);
    CGImageDestinationAddImage(destination, _image, (__bridge CFDictionaryRef) properties);
    CGImageDestinationFinalize(destination);
    
    CFRelease(type);
    CFRelease(destination);
    
    return (__bridge_transfer NSData *)imageData;
}

- (NSInteger)getWidth {
    return CGImageGetWidth(_image);
}

- (NSInteger)getHeight {
    return CGImageGetHeight(_image);
}

- (void)crop:(CGRect)cropRectangle {
    CGImageRef newImage = CGImageCreateWithImageInRect(_image, cropRectangle);
    CGImageRelease(_image);
    _image = newImage;
}

- (void)drawBorder:(NSString*)color
       withLeftTop:(CGSize)leftTop
   withRightBottom:(CGSize)rightBottom {
    double orgWidth = [self getWidth];
    double orgHeight = [self getHeight];
    double newWidth = orgWidth + leftTop.width + rightBottom.width;
    double newHeight = orgHeight + leftTop.height + rightBottom.height;
    
    CGColorSpaceRef colorSpace = CGImageGetColorSpace(_image);
    CGContextRef context = CGBitmapContextCreate(nil,
                                             newWidth,
                                             newHeight,
                                             CGImageGetBitsPerComponent(_image),
                                             32 * newWidth,
                                             colorSpace,
                                             CGImageGetAlphaInfo(_image));
    CGContextBeginPath(context);
    CGColorRef borderColor = CGColorRetain([UIColor parseColor:color].CGColor);
    CGContextSetFillColorWithColor(context, borderColor);
    CGContextFillRect(context, CGRectMake(0, 0, newWidth, newHeight));
    
    CGContextDrawImage(context, CGRectMake(leftTop.width, rightBottom.height, orgWidth, orgHeight), _image);
    
    CGImageRef newImage = CGBitmapContextCreateImage(context);
    
    CGImageRelease(_image);
    CGColorRelease(borderColor);
    CGContextRelease(context);
    
    _image = newImage;
}

- (void)dealloc {
    _contextId = nil;
    if (_image) {
        CGImageRelease(_image);
        _image = nil;
    }
}

@end

File: ios/RCTIONImage2DContext.swift
//
//  RCTIONImage2DContext.swift
//  RCTIONImage2D
//
//  Created by Marko on 14/01/16.
//  
//

import Foundation
import ImageIO
import MobileCoreServices

class RCTIONImage2DContext: NSObject {
  private var contextId = NSUUID().UUIDString
  private var image: CGImageRef?
  
  func getContextId() -> String {
    return self.contextId;
  }
  
  func createFromFileUrl(fileUrl: String, withMaxWidth maxWidth: Int, withMaxHeight maxHeight: Int) -> Void {
    let imageData = NSData.init(contentsOfURL: NSURL.init(fileURLWithPath: fileUrl))!
    self.createFromData(imageData, withMaxWidth: maxWidth, withMaxHeight: maxHeight)
  }
  
  func createFromBase64String(base64: String, withMaxWidth maxWidth: Int, withMaxHeight maxHeight: Int) -> Void {
    let imageData = NSData.init(base64EncodedString: base64, options: NSDataBase64DecodingOptions(rawValue: 0))!
    self.createFromData(imageData, withMaxWidth: maxWidth, withMaxHeight: maxHeight)
  }
  
  private func createFromData(imageData: NSData, withMaxWidth maxWidth: Int, withMaxHeight maxHeight: Int) -> Void {
    let imageSource = CGImageSourceCreateWithData(imageData, nil)!
    let newImage = CGImageSourceCreateImageAtIndex(imageSource, 0, nil)!
    self.createFromImage(newImage, withMaxWidth: maxWidth, withMaxHeight: maxHeight)
  }
  
  private func createFromImage(image: CGImageRef, withMaxWidth maxWidth: Int, withMaxHeight maxHeight: Int) -> Void {
    let orgWidth = Double(CGImageGetWidth(image))
    let orgHeight = Double(CGImageGetHeight(image))
    var calcNewWidth = Double(maxWidth)
    var calcNewHeight = Double(maxHeight)
    
    if orgWidth <= calcNewWidth && orgHeight <= calcNewHeight {
      self.image = image
      return;
    }
    
    if orgWidth / calcNewWidth < orgHeight / calcNewHeight {
      calcNewWidth = (calcNewHeight / orgHeight) * orgWidth
    } else {
      calcNewHeight = (calcNewWidth / orgWidth) * orgHeight
    }
    
    let newWidth = Int(calcNewWidth)
    let newHeight = Int(calcNewHeight)
    
    let context = CGBitmapContextCreate(
      nil,
      newWidth,
      newHeight,
      CGImageGetBitsPerComponent(image),
      32 * newWidth,
      CGImageGetColorSpace(image),
      CGImageGetAlphaInfo(image).rawValue
    )
    
    CGContextDrawImage(context, CGRectMake(0, 0, CGFloat(newWidth), CGFloat(newHeight)), image)
    
    self.image = CGBitmapContextCreateImage(context)
  }
  
  func save(fileName: String) -> String {
    let imageData = self.getAsData()
    
    let documentsDirectory: NSString = NSSearchPathForDirectoriesInDomains(.DocumentDirectory, .UserDomainMask, true)[0]
    let fileManager = NSFileManager.defaultManager()
    
    let imagePath = (documentsDirectory.stringByAppendingPathComponent(fileName) as NSString).stringByAppendingPathExtension("jpg")!
    fileManager.createFileAtPath(imagePath, contents: imageData, attributes: nil)
    
    return imagePath
  }
  
  func getAsBase64String() -> String {
    return self.getAsData().base64EncodedStringWithOptions(NSDataBase64EncodingOptions(rawValue: 0))
  }
  
  private func getAsData() -> NSData {
    let imageData = CFDataCreateMutable(nil, 0)
    let type = UTTypeCreatePreferredIdentifierForTag(kUTTagClassMIMEType, "image/png", kUTTypeImage)!.takeRetainedValue()
    let destination = CGImageDestinationCreateWithData(imageData, type, 1, nil)!
    CGImageDestinationAddImage(destination, self.image!, nil)
    CGImageDestinationFinalize(destination)
    
    return imageData
  }
  
  func getWidth() -> Int {
    return CGImageGetWidth(self.image);
  }
  
  func getHeight() -> Int {
    return CGImageGetHeight(self.image);
  }
  
  func crop(cropRectangle: CGRect) -> Void {
    self.image = CGImageCreateWithImageInRect(self.image, cropRectangle);
  }
  
  func drawBorder(color: String, withLeftTop leftTop: CGSize, withRightBottom rightBottom: CGSize) throws -> Void {
    let orgWidth = self.getWidth()
    let orgHeight = self.getHeight()
    let newWidth = orgWidth + Int(leftTop.width + rightBottom.width);
    let newHeight = orgHeight + Int(leftTop.height + rightBottom.height);
    
    let context = CGBitmapContextCreate(
      nil,
      newWidth,
      newHeight,
      CGImageGetBitsPerComponent(self.image),
      32 * newWidth,
      CGImageGetColorSpace(self.image),
      CGImageGetAlphaInfo(self.image).rawValue
    )
    
    CGContextBeginPath(context);
    CGContextSetFillColorWithColor(context, try UIColor.parseColor(color).CGColor);
    CGContextFillRect(context, CGRectMake(0, 0, CGFloat(newWidth), CGFloat(newHeight)));
    CGContextDrawImage(context, CGRectMake(leftTop.width, rightBottom.height, CGFloat(orgWidth), CGFloat(orgHeight)), self.image);
    
    self.image = CGBitmapContextCreateImage(context);
  }
  
  deinit {
    self.image = nil;
  }
}
